sympy/matrices/eigen.py

```diff
@@ -0,0 +1,998 @@
+        a vector of length 3 is returned as ``Matrix([a_1, a_2, a_3])``.
+
+    Raises
+    ======
+
+    NotImplementedError
+        If failed to compute nullspace.
+
+    Examples
+    ========
+
+    >>> from sympy.matrices import Matrix
+    >>> M = Matrix(3, 3, [0, 1, 1, 1, 0, 0, 1, 1, 1])
+    >>> M.eigenvects()
+    [(-1, 1, [Matrix([
+    [-1],
+    [ 1],
+    [ 0]])]), (0, 1, [Matrix([
+    [ 0],
+    [-1],
+    [ 1]])]), (2, 1, [Matrix([
+    [2/3],
+    [1/3],
+    [  1]])])]
+
+    See Also
+    ========
+
+    eigenvals
+    MatrixSubspaces.nullspace
+    """
+
+    def eigenspace(eigenval):
+        """Get a basis for the eigenspace for a particular eigenvalue"""
+
+        m   = M - M.eye(M.rows) * eigenval
+        ret = m.nullspace(iszerofunc=iszerofunc, dotprodsimp=dotprodsimp)
+
+        # the nullspace for a real eigenvalue should be
+        # non-trivial.  If we didn't find an eigenvector, try once
+        # more a little harder
+        if len(ret) == 0 and simplify:
+            ret = m.nullspace(iszerofunc=iszerofunc, simplify=True, dotprodsimp=dotprodsimp)
+        if len(ret) == 0:
+            raise NotImplementedError(
+                    "Can't evaluate eigenvector for eigenvalue %s" % eigenval)
+
+        return ret
+
+    simplify = flags.get('simplify', True)
+
+    if not isinstance(simplify, FunctionType):
+        simpfunc = _simplify if simplify else lambda x: x
+
+    primitive = flags.get('simplify', False)
+    chop      = flags.pop('chop', False)
+
+    flags.pop('multiple', None)  # remove this if it's there
+
+    has_floats = M.has(Float) # roots doesn't like Floats, so replace them with Rationals
+
+    if has_floats:
+        M = M.applyfunc(lambda x: nsimplify(x, rational=True))
+
+    eigenvals = M.eigenvals(rational=False,
+            error_when_incomplete=error_when_incomplete,
+            dotprodsimp=dotprodsimp, **flags)
+
+    ret = [(val, mult, eigenspace(val)) for val, mult in
+                sorted(eigenvals.items(), key=default_sort_key)]
+
+    if primitive:
+        # if the primitive flag is set, get rid of any common
+        # integer denominators
+        def denom_clean(l):
+            from sympy import gcd
+            return [(v / gcd(list(v))).applyfunc(simpfunc) for v in l]
+
+        ret = [(val, mult, denom_clean(es)) for val, mult, es in ret]
+
+    if has_floats:
+        # if we had floats to start with, turn the eigenvectors to floats
+        ret = [(val.evalf(chop=chop), mult, [v.evalf(chop=chop) for v in es])
+                for val, mult, es in ret]
+
+    return ret
+
+
+def _is_diagonalizable_with_eigen(M, reals_only=False, dotprodsimp=None):
+    """See _is_diagonalizable. This function returns the bool along with the
+    eigenvectors to avoid calculating them again in functions like
+    ``diagonalize``."""
+
+    if not M.is_square:
+        return False, []
+
+    eigenvecs = M.eigenvects(simplify=True, dotprodsimp=dotprodsimp)

```

Comment:
Should thsi not use `_eigenvects` from above?

Reply:
In both of these cases it could, I did not really dive into these kind of changes to avoid modifying too much code while relocating (which is the most important thing, afterwards the functions can be tweaked). I have a list of issues to be sorted after the relocation is done and will add this to that list - where applicable since in some cases maybe we want to use methods where it is possible they will be overloaded by some type of matrix?